ARROW-10643: [Python] Pandas<->pyarrow roundtrip failing to recreate index for empty dataframe

AlenkaF opened the pull request:
This PR tries to correct the roundtrip of an empty `pandas.DataFrame` with `RangeIndex` (so no columns, but a non-zero shape for the rows) by adding a check for empty columns and a `pandas.RangeIndex` in the `from_arrays` method called from `from_pandas` and then creating an empty table with schema and `num_rows`.

jorisvandenbossche commented on python/pyarrow/tests/test_table.py:
```diff
@@ -339,23 +339,29 @@ def test_chunked_array_to_pandas_preserve_name():
+    assert table.num_rows == 4
+    assert data.shape == (4, 0)
+    assert result.shape == (4, 0)
+
```
Maybe add a `assert result.index.equals(data.index)`, if that passes

jorisvandenbossche commented on python/pyarrow/table.pxi:
```diff
@@ -1771,6 +1771,22 @@ cdef class Table(_PandasConvertible):
+                    result = pyarrow_wrap_table(
+                        CTable.MakeWithRows(c_schema, columns, n_rows))
+            except IndexError:
+                pass
```
I would maybe move this check to the `from_pandas` method, as it is specific to pandas input, and not something you should ever get from directly using `from_arrays`

jorisvandenbossche replied:
Or we could maybe even move the actual check to `dataframe_to_arrays`, since you already have a `pandas_metadata` object there. And then we can add a third return argument that is always None, except if you have this case of no columns but a range index (and then can return the range index info or length)

AlenkaF replied:
Good idea! Agree 💯 
Will try it out now and see if I can make it work.

jorisvandenbossche commented on python/pyarrow/table.pxi:
```diff
@@ -1771,6 +1771,22 @@ cdef class Table(_PandasConvertible):
+
+        # In case of an empty dataframe with RangeIndex -> create an empty Table with
+        # number of rows equal to Index length
+        if arrays == [] and schema is not None:
```
```suggestion
        if len(arrays) == 0 and schema is not None:
```

AlenkaF commented:
I have split the code, hope it makes sense.
I also added the same logic for RecordBatch as they are both using `dataframe_to_arrays` in `from_pandas`.

jorisvandenbossche commented on python/pyarrow/pandas_compat.py:
```diff
@@ -623,7 +623,21 @@ def _can_definitely_zero_copy(arr):
+    # If dataframe is empty but with RangeIndex ->
+    # remember the length of the indexes
+    n_rows = None
+    if len(arrays) == 0 and schema is not None:
```
Is the `schema is not None` needed? Above, there is a block `if schema is None` which then creates a schema, which seems to indicate that at this point a `schema` object will always be `not None`?

AlenkaF replied:
Oh yes, you are right. Removed it.

jorisvandenbossche commented on python/pyarrow/pandas_compat.py:
```diff
@@ -623,7 +623,21 @@ def _can_definitely_zero_copy(arr):
+                start = index_descriptors[0]["start"]
+                stop = index_descriptors[0]["stop"]
+                step = index_descriptors[0]["step"]
+                n_rows = (stop - start - 1)//step + 1
```
Is the `- 1` part needed? (in any case for the test example of (0, 10, 3) it's not needed)

And I am just thinking: a `step` can be negative, so we should take that into account (or at least test, it might already work because `stop - start` will then also be negative, so the result would still be positive).

jorisvandenbossche replied:
Actually, even simpler is to do `n_rows = len(range(start, stop, step))` and let python calculate the length of the range for us!

AlenkaF replied:
I am just sorry I haven't found this myself :)

jorisvandenbossche commented on python/pyarrow/table.pxi:
```diff
@@ -1177,13 +1177,13 @@ cdef class RecordBatch(_PandasConvertible):
 
     @staticmethod
-    def from_arrays(list arrays, names=None, schema=None, metadata=None):
+    def from_arrays(list arrays, names=None, schema=None, metadata=None, n_rows=None):
```
If we keep the keyword, I would also make this `num_rows` to keep it consistent with the public property of a Table. 
But, another option is also to handle this `n_rows==0` case in `from_pandas`. Now, I suppose in theory it can be useful to expose here, to make it possible to create an empty table with a given number of rows via the Python API (although is is very much an esoteric use case)

AlenkaF replied:
To sum up the conclusion from our talk today: there would be added confusion and not much value in keeping `n_row` exposed in `from_arrays`. I will move the code to `from_pandas`.

I am figuring out the Cython code and will commit as soon as I am finished.

jorisvandenbossche reviewed (commented):
Thanks for the updates!

jorisvandenbossche commented on python/pyarrow/table.pxi:
```diff
@@ -1779,15 +1796,41 @@ cdef class Table(_PandasConvertible):
+        if n_rows:
+            names = None
+            metadata = None
+            converted_arrays = _sanitize_arrays(arrays, names, schema, metadata,
+                                                &c_schema)
+            arr.reserve(0)
+            for item in converted_arrays:
+                if isinstance(item, Array):
+                    arr.push_back(
+                        make_shared[CChunkedArray](
+                            (<Array> item).sp_array
+                        )
+                    )
+                elif isinstance(item, ChunkedArray):
+                    arr.push_back((<ChunkedArray> item).sp_chunked_array)
+                else:
+                    raise TypeError(type(item))
```
I think this can maybe be simplified? Because we now that `n_rows` is only defined if `arrays` is empty? In which case this for loop would not do anything? 
And this also means that we don't need to call `_sanitize_arrays` I think. You can pass the existing schema (returned from `dataframe_to_arrays`) to `CTable.MakeWithRows` (to get the CSchema, you can unwrap it: `schema.unwrap()`)

AlenkaF replied:
I wasn't able to build without cython erroring if I didn't use this whole part - but definitely agree, it should be simplified somehow. I'll try what you are proposing, thanks!

AlenkaF replied:
Creating an empty cython vector and converting schema to cython is simplified now, thank you for the live help!

jorisvandenbossche commented on python/pyarrow/table.pxi:
```diff
@@ -1177,10 +1177,18 @@ cdef class RecordBatch(_PandasConvertible):
-        return cls.from_arrays(arrays, schema=schema)
+
+        # If df is empty but row index is not, create empty RecordBatch with rows >0
+        cdef:
+            vector[shared_ptr[CArray]] c_arrays
```
```suggestion
        cdef vector[shared_ptr[CArray]] c_arrays
```

(for a single definition can also put it on a single line)

jorisvandenbossche commented on python/pyarrow/table.pxi:
```diff
@@ -1779,15 +1787,23 @@ cdef class Table(_PandasConvertible):
-        return cls.from_arrays(arrays, schema=schema)
+
+        # If df is empty but row index is not, create empty Table with rows >0
+        cdef:
+            vector[shared_ptr[CChunkedArray]] c_arrays
```
```suggestion
        cdef vector[shared_ptr[CChunkedArray]] c_arrays
```

jorisvandenbossche reviewed (approved):
Thank!
